File: launcher.py
"""
MAQ AUTO EDITOR ULTRA - Universal Python Launcher
Boots local backend, checks ports, and opens the default browser UI automatically.
"""

import os
import sys
import webbrowser
import threading
import time
import urllib.request
import subprocess

PORT = int(os.environ.get('PORT', 4000))
URL = f"http://localhost:{PORT}"

def check_and_open():
    for _ in range(30):
        time.sleep(0.5)
        try:
            with urllib.request.urlopen(URL, timeout=1) as response:
                if response.status == 200:
                    print(f"\n[OK] Backend server is running at {URL}")
                    print(f"[OK] Opening browser automatically...\n")
                    webbrowser.open(URL)
                    break
        except Exception:
            pass

if __name__ == '__main__':
    print("===========================================================")
    print("       MAQ AUTO EDITOR ULTRA - DESKTOP LAUNCHER           ")
    print("===========================================================")
    print(f"Starting server on {URL}...")

    # Start browser opener in background thread
    threading.Thread(target=check_and_open, daemon=True).start()

    # Import and run server
    from backend.server import run_server
    run_server(PORT)

File: launcher.js
/**
 * MAQ AUTO EDITOR ULTRA - Local Application Launcher
 * Boots the local backend server, checks port availability, and automatically launches the desktop browser UI.
 */

const { spawn, exec } = require('child_process');
const http = require('http');
const path = require('path');
const os = require('os');

const PORT = process.env.PORT || 4000;
const URL = `http://localhost:${PORT}`;

console.log('===========================================================');
console.log('       MAQ AUTO EDITOR ULTRA - LOCAL DESKTOP LAUNCHER      ');
console.log('===========================================================');
console.log(`[1/3] Starting backend engine on ${URL}...`);

// Start backend server
require('./backend/server.js');

function openBrowser(targetUrl) {
  const platform = process.platform;
  let cmd = '';

  if (platform === 'win32') {
    cmd = `start "" "${targetUrl}"`;
  } else if (platform === 'darwin') {
    cmd = `open "${targetUrl}"`;
  } else {
    cmd = `xdg-open "${targetUrl}" 2>/dev/null || sensible-browser "${targetUrl}" 2>/dev/null || true`;
  }

  exec(cmd, (err) => {
    if (err) {
      console.log(`Please open your browser manually at: ${targetUrl}`);
    } else {
      console.log(`[3/3] Desktop UI opened successfully in browser: ${targetUrl}`);
    }
  });
}

// Poll server until ready, then open browser
let attempts = 0;
const checkInterval = setInterval(() => {
  attempts++;
  http.get(URL, (res) => {
    if (res.statusCode === 200) {
      clearInterval(checkInterval);
      console.log(`[2/3] Backend server is healthy (HTTP 200 OK).`);
      openBrowser(URL);
    }
  }).on('error', () => {
    if (attempts > 30) {
      clearInterval(checkInterval);
      console.error('Server startup timed out.');
    }
  });
}, 400);
